### ISSUE TITLE
Legacy debugAdapter "can not run under Rosetta"

### ISSUE THREAD

**Issue by shayneo** [golang/vscode-go#1936]
### What version of Go, VS Code & VS Code Go extension are you using?
<details><summary>Version Information</summary><br>

* Run `go version` to get version of Go from _the VS Code integrated terminal_.
	- go1.17.5 darwin/arm64
* Run `gopls -v version` to get version of Gopls from _the VS Code integrated terminal_.
```
golang.org/x/tools/gopls v0.7.3
    golang.org/x/tools/gopls@v0.7.3 h1:Lru57ht8vtDMouRskFC085VAjBAZRAISd/lwvwOOV0Q=
    github.com/BurntSushi/toml@v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=
    github.com/google/go-cmp@v0.5.6 h1:BKbKCqvP6I+rmFHt06ZmyQtvB8xAkWdhFyr0ZUNZcxQ=
    github.com/sergi/go-diff@v1.1.0 h1:we8PVUC3FE2uYfodKH/nBHMSetSfHDR6scGdBi+erh0=
    golang.org/x/mod@v0.4.2 h1:Gz96sIWK3OalVv/I/qNygP42zyoKp3xptRVCWRFEBvo=
    golang.org/x/sync@v0.0.0-20210220032951-036812b2e83c h1:5KslGYwFpkhGh+Q16bwMP3cOontH8FOep7tGV86Y7SQ=
    golang.org/x/sys@v0.0.0-20210809222454-d867a43fc93e h1:WUoyKPm6nCo1BnNUvPGnFG3T5DUVem42yDJZZ4CNxMA=
    golang.org/x/text@v0.3.6 h1:aRYxNxv6iGQlyVaZmk6ZgYEDa+Jg18DxebPSrd6bg1M=
    golang.org/x/tools@v0.1.8-0.20211014194737-fc98fb2abd48 h1:hk7xRoeg0CG1nRLsd5BZLDUgVpA9bnKylGk1p2/BPH0=
    golang.org/x/xerrors@v0.0.0-20200804184101-5ec99f83aff1 h1:go1bK/D/BFZV2I8cIQd1NKEZ+0owSTG1fDTci4IqFcE=
    honnef.co/go/tools@v0.2.0 h1:ws8AfbgTX3oIczLPNPCu5166oBg9ST2vNs0rcht+mDE=
    mvdan.cc/gofumpt@v0.1.1 h1:bi/1aS/5W00E2ny5q65w9SnKpWEF/UIOqDYBILpo9rA=
    mvdan.cc/xurls/v2@v2.3.0 h1:59Olnbt67UKpxF1EwVBopJvkSUBmgtb468E4GVWIZ1I=
```
* Run `code -v` or `code-insiders -v` to get version of VS Code or VS Code Insiders.
```
1.63.0
7db1a2b88f7557e0a43fec75b6ba7e50b3e9f77e
arm64
```
* Check your installed extensions to get the version of the VS Code Go extension
	v0.29.0
* Run Ctrl+Shift+P (Cmd+Shift+P on Mac OS) > `Go: Locate Configured Go Tools` command.
```
Tools environment: GOPATH=/Users/shayneosullivan/projects/vinli/go
Installing 10 tools at /Users/shayneosullivan/projects/vinli/go/bin in module mode.
  gopkgs
  go-outline
  gotests
  gomodifytags
  impl
  goplay
  dlv
  dlv-dap
  staticcheck
  gopls

Installing github.com/uudashr/gopkgs/v2/cmd/gopkgs@latest (/Users/shayneosullivan/projects/vinli/go/bin/gopkgs) SUCCEEDED
Installing github.com/ramya-rao-a/go-outline@latest (/Users/shayneosullivan/projects/vinli/go/bin/go-outline) SUCCEEDED
Installing github.com/cweill/gotests/gotests@latest (/Users/shayneosullivan/projects/vinli/go/bin/gotests) SUCCEEDED
Installing github.com/fatih/gomodifytags@latest (/Users/shayneosullivan/projects/vinli/go/bin/gomodifytags) SUCCEEDED
Installing github.com/josharian/impl@latest (/Users/shayneosullivan/projects/vinli/go/bin/impl) SUCCEEDED
Installing github.com/haya14busa/goplay/cmd/goplay@latest (/Users/shayneosullivan/projects/vinli/go/bin/goplay) SUCCEEDED
Installing github.com/go-delve/delve/cmd/dlv@latest (/usr/local/bin/dlv) SUCCEEDED
Installing github.com/go-delve/delve/cmd/dlv@master (/Users/shayneosullivan/projects/vinli/go/bin/dlv-dap) SUCCEEDED
Installing honnef.co/go/tools/cmd/staticcheck@latest (/Users/shayneosullivan/projects/vinli/go/bin/staticcheck) SUCCEEDED
Installing golang.org/x/tools/gopls@latest (/Users/shayneosullivan/projects/vinli/go/bin/gopls) SUCCEEDED

All tools successfully installed. You are ready to Go. :)
```
</details>

### Share the Go related settings you have added/edited
```
 "go.toolsManagement.autoUpdate" : true,
 "[go]" : {
    "editor.defaultFormatter" : "golang.go"
  },
  "go.delveConfig" : {
    "useApiV1" : false,
    "debugAdapter" : "legacy",
    "dlvLoadConfig" : {
      "maxStringLen" : 4000,
      "maxStructFields" : -1,
      "followPointers" : true,
      "maxArrayValues" : 400,
      "maxVariableRecurse" : 3,
    }
  },
```

### Describe the bug
Running the debugger on an M1 mac with `"debugAdapter": "legacy"` produces the following error:
```
API server listening at: 127.0.0.1:22360
could not launch process: can not run under Rosetta, check that the installed build of Go is right for your CPU architecture
Process exiting with code: 1
```

### Steps to reproduce the behavior:
1. Set debugAdapter to legacy
2. Run debugger in test mode
3. See error

**Comment by findleyr** [comment]
It looks like delve is not finding the darwin/arm64 version of Go. See also #1669.

Do you have any other versions of Go on your system? Assuming vscode is using the correct version, please try running `Go: install/update tools`.

**Comment by shayneo** [comment]
I've run `Go: install/update tools`, which didn't help.

I only have 1 version of Go on my machine. 

What's weird is that I **don't** get this error when running the debugger with the default debugAdapter (`"dlv-dap"`)

**Comment by findleyr** [comment]
@suzmue is it possible that the legacy debug adapter is using a different version of `go`, or of `dlv`?

Aside: is there a reason you can't use `dlv-dap`?

**Comment by findleyr** [comment]
Also, in addition to `whereis go`, can you run `whereis dlv`? Do you have more than one `dlv` binary?

**Comment by shayneo** [comment]
I'm also suspicious that the legacy adapter is using a different go/dlv

```
which go
/usr/local/go/bin/go
```

```
which dlv
/usr/local/bin/dlv
```

**Comment by shayneo** [comment]
Re using `dlv-dap`, I'm all for that approach. I've been trying to get that working with our setup... we use [ginkgo](https://github.com/onsi/ginkgo) and can't seem to get that working with `dlv-dap`. I'm going to keep tinkering, but thought it was worth reporting on the potential issue with `legacy`

**Comment by findleyr** [comment]
> which dlv

Can you use `whereis`? `which` just returns one result?

Can you also please run, _from the vscode integrated terminal_:
 - `echo $GOBIN`
 - `go env GOPATH`

**Comment by shayneo** [comment]
I get no results when using `whereis dlv`

`echo $GOBIN` - returns nothing

```
go env GOPATH
/Users/shayneosullivan/projects/vinli/go
```

**Comment by findleyr** [comment]
> go env GOPATH
> /Users/shayneosullivan/projects/vinli/go

Do you have a dlv in /Users/shayneosullivan/projects/vinli/go/bin?

**Comment by suzmue** [comment]
Both dlv and dlv-dap should be using the same version of Go, but they are likely using a different version of delve (dlv is a stable version, and dlv-dap will build from tip). This is a temporary workaround that was added because the DAP server was under rapid development. 

If you want to use the dlv-dap binary with the legacy adapter, you can do so by setting
```
    "go.alternateTools": {
        "dlv": "/path/to/dlv-dap"
    }
```

You can also try updating the dlv command by using `> Go: Install / Update Tools` from the command palette.

**Comment by findleyr** [comment]
> You can also try updating the dlv command by using > Go: Install / Update Tools from the command palette.

@suzmue this was already tried. Is it possible that there's a bug with the logic to maintain two `dlv` installs, such that we're not rebuilding the `dlv` that @shayneo is encountering? (I suspect that the problematic dlv is in `$GOPATH/bin`).

@shayneo I should have asked this earlier, but in the past had you installed a non-arm version of Go?

**Comment by shayneo** [comment]
> > go env GOPATH
> > /Users/shayneosullivan/projects/vinli/go
> 
> Do you have a dlv in /Users/shayneosullivan/projects/vinli/go/bin?

I do indeed! it's the exact version as `/usr/local/bin/dlv`

**Comment by shayneo** [comment]
> > You can also try updating the dlv command by using > Go: Install / Update Tools from the command palette.
> 
> @suzmue this was already tried. Is it possible that there's a bug with the logic to maintain two `dlv` installs, such that we're not rebuilding the `dlv` that @shayneo is encountering? (I suspect that the problematic dlv is in `$GOPATH/bin`).
> 
> @shayneo I should have asked this earlier, but in the past had you installed a non-arm version of Go?

Yes. I got a new m1 machine earlier this week and migrated my intel mac, so the amd go version was brought along in the migration. I think I've completely purged all other go versions though, so I should only have go1.17.5 darwin/arm64 at this point.

**Comment by findleyr** [comment]
Hmm, can you just try the following from the vscode terminal?

```
go install github.com/go-delve/delve/cmd/dlv@latest
```

Beyond this, I'm not sure how much we can help you. Somewhere something is not arm64 :/  It sounds like you should purge/rebuild your $GOPATH/bin, if binaries there were potentially migrated from your intel mac. Please do let us know what you discover, as it may indicate a flaw in our update logic.

**Comment by shayneo** [comment]
Got it to work! 🎉 

Here's how:
1. Manually removed everything from my `$GOPATH/bin`
2. Ran `> Go: Install / Update Tools`
3. Once I did this, I started getting a new error (`could not launch process: EOF`). To fix that, had to remove and install xcode command line tools, as covered here: https://github.com/go-delve/delve/issues/2794#issuecomment-989797645

Thanks @findleyr  for the help and quick responses!
